plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/JdbcMergeSink.java

```diff
@@ -0,0 +1,322 @@
+import io.trino.spi.connector.ColumnHandle;
+import io.trino.spi.connector.ConnectorMergeSink;
+import io.trino.spi.connector.ConnectorMergeTableHandle;
+import io.trino.spi.connector.ConnectorPageSink;
+import io.trino.spi.connector.ConnectorPageSinkId;
+import io.trino.spi.connector.ConnectorSession;
+import io.trino.spi.predicate.Domain;
+import io.trino.spi.predicate.TupleDomain;
+import io.trino.spi.type.Type;
+
+import java.sql.Connection;
+import java.sql.SQLException;
+import java.util.Collection;
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+import java.util.concurrent.CompletableFuture;
+import java.util.stream.IntStream;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.collect.ImmutableSet.toImmutableSet;
+import static io.trino.plugin.jdbc.JdbcErrorCode.JDBC_ERROR;
+import static io.trino.spi.type.BigintType.BIGINT;
+import static io.trino.spi.type.TinyintType.TINYINT;
+import static java.util.Objects.requireNonNull;
+import static java.util.concurrent.CompletableFuture.completedFuture;
+
+public class JdbcMergeSink
+        implements ConnectorMergeSink
+{
+    private final int columnCount;
+    private final int[] updateAssignmentChannels;
+    private final ConnectorPageSinkId pageSinkId;
+
+    private final ConnectorPageSink insertSink;
+    private final ConnectorPageSink updateSink;
+    private final ConnectorPageSink deleteSink;
+
+    public JdbcMergeSink(
+            ConnectorSession session,
+            ConnectorMergeTableHandle mergeTableHandle,
+            JdbcClient jdbcClient,
+            ConnectorPageSinkId pageSinkId,
+            RemoteQueryModifier queryModifier,
+            QueryBuilder queryBuilder)
+    {
+        requireNonNull(session, "session is null");
+        requireNonNull(mergeTableHandle, "mergeTableHandle is null");
+        requireNonNull(jdbcClient, "jdbcClient is null");
+        requireNonNull(queryModifier, "queryModifier is null");
+        requireNonNull(queryBuilder, "queryBuilder is null");
+
+        JdbcMergeTableHandle mergeHandle = (JdbcMergeTableHandle) mergeTableHandle;
+        JdbcOutputTableHandle outputHandle = mergeHandle.getOutputTableHandle();
+        List<JdbcColumnHandle> primaryKeys = mergeHandle.getPrimaryKeys();
+        checkArgument(!primaryKeys.isEmpty(), "primary keys not exists");
+
+        this.columnCount = outputHandle.getColumnNames().size();
+
+        this.updateAssignmentChannels = getUpdateAssignmentChannels(outputHandle, primaryKeys);
+
+        this.pageSinkId = requireNonNull(pageSinkId, "pageSinkId is null");
+
+        // The TupleDomain for building the conjuncts of the primary keys
+        ImmutableMap.Builder<ColumnHandle, Domain> primaryKeysDomainBuilder = ImmutableMap.builder();
+        // This value is to build the TupleDomain, but it won't affect the query field in result of the `QueryBuilder#prepareDeleteQuery` and `QueryBuilder#prepareUpdateQuery`
```

Comment:
This comment needs a rewrite. Let's write that value won't be used as the real value will be passed to `storeMergedRows`